File: src/main/python/AoC2023_01.py
#! /usr/bin/env python3
#
# Advent of Code 2023 Day 1
#

import sys
from collections.abc import Callable

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples

Input = InputData
Output1 = int
Output2 = int

TEST1 = """\
1abc2
pqr3stu8vwx
a1b2c3d4e5f
treb7uchet
"""
TEST2 = """\
two1nine
eightwothree
abcone2threexyz
xtwone3four
4nineeightseven2
zoneight234
7pqrstsixteen
"""


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return input_data

    def _solve(self, lines: InputData, f: Callable[[str], list[int]]) -> int:
        return sum(digits[0] * 10 + digits[-1] for digits in map(f, lines))

    def part_1(self, lines: InputData) -> Output1:
        def get_digits(line: str) -> list[int]:
            return [int(c) for c in line if c.isdigit()]

        return self._solve(lines, get_digits)

    def part_2(self, lines: InputData) -> Output2:
        def get_digits(line: str) -> list[int]:
            nums = [
                "one",
                "two",
                "three",
                "four",
                "five",
                "six",
                "seven",
                "eight",
                "nine",
            ]

            def find_digit(s: str) -> int | None:
                if s[0].isdigit():
                    return int(s[0])
                for j, num in enumerate(nums):
                    if s.startswith(num):
                        return j + 1
                return None

            return [
                x
                for x in (find_digit(line[i:]) for i in range(len(line)))
                if x is not None
            ]

        return self._solve(lines, get_digits)

    @aoc_samples(
        (
            ("part_1", TEST1, 142),
            ("part_2", TEST2, 281),
        )
    )
    def samples(self) -> None:
        pass


solution = Solution(2023, 1)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()

File: src/main/java/AoC2023_01.java
import static com.github.pareronia.aoc.IntegerSequence.Range.range;

import com.github.pareronia.aoc.Utils;
import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

import java.util.List;
import java.util.Objects;
import java.util.function.Function;

@SuppressWarnings({"PMD.ClassNamingConventions", "PMD.NoPackage"})
public final class AoC2023_01 extends SolutionBase<List<String>, Integer, Integer> {

    private AoC2023_01(final boolean debug) {
        super(debug);
    }

    public static AoC2023_01 create() {
        return new AoC2023_01(false);
    }

    public static AoC2023_01 createDebug() {
        return new AoC2023_01(true);
    }

    @Override
    protected List<String> parseInput(final List<String> inputs) {
        return inputs;
    }

    private int solve(final List<String> input, final Function<String, List<Integer>> getDigits) {
        return input.stream()
                .map(getDigits)
                .mapToInt(digits -> digits.get(0) * 10 + digits.getLast())
                .sum();
    }

    @Override
    public Integer solvePart1(final List<String> input) {
        final Function<String, List<Integer>> getDigits =
                line ->
                        Utils.asCharacterStream(line)
                                .filter(Character::isDigit)
                                .map(c -> Integer.parseInt(String.valueOf(c)))
                                .toList();
        return solve(input, getDigits);
    }

    @Override
    public Integer solvePart2(final List<String> input) {
        final List<String> nums =
                List.of("one", "two", "three", "four", "five", "six", "seven", "eight", "nine");
        final Function<String, Integer> findDigit =
                s -> {
                    if (Character.isDigit(s.charAt(0))) {
                        return Integer.parseInt(String.valueOf(s.charAt(0)));
                    } else {
                        for (int j = 0; j < nums.size(); j++) {
                            if (s.startsWith(nums.get(j))) {
                                return j + 1;
                            }
                        }
                    }
                    return null;
                };
        final Function<String, List<Integer>> getDigits =
                line ->
                        range(line.length()).stream()
                                .map(line::substring)
                                .map(findDigit)
                                .filter(Objects::nonNull)
                                .toList();
        return solve(input, getDigits);
    }

    @Samples({
        @Sample(method = "part1", input = TEST1, expected = "142"),
        @Sample(method = "part2", input = TEST2, expected = "281"),
    })
    public static void main(final String[] args) throws Exception {
        create().run();
    }

    private static final String TEST1 =
            """
            1abc2
            pqr3stu8vwx
            a1b2c3d4e5f
            treb7uchet
            """;

    private static final String TEST2 =
            """
            two1nine
            eightwothree
            abcone2threexyz
            xtwone3four
            4nineeightseven2
            zoneight234
            7pqrstsixteen
            """;
}
